src/compiler-jdk11/scala/tools/nsc/jpms/JpmsClassPath.scala

```diff
@@ -0,0 +1,219 @@
+
+  private val fileManager = impl.getFileManager
+  private def locationAsPaths(location: StandardLocation): IndexedSeq[Path] = getLocationAsPaths(fileManager, location).asScala.toVector
+  private val packageIndex = mutable.AnyRefMap[String, PackageEntryImpl]()
+  private val moduleLocations = Array(StandardLocation.MODULE_SOURCE_PATH, StandardLocation.UPGRADE_MODULE_PATH, StandardLocation.SYSTEM_MODULES, StandardLocation.MODULE_PATH, StandardLocation.PATCH_MODULE_PATH)
+  private val paths = mutable.ArrayBuffer[(Path, StandardLocation, Location)]()
+  indexLocations()
+
+  // TODO JPMS It's a bit annoying that we have to start interrogating the classpath before we've resolve the module
+  //      graph: we have to assume that all modules are potentially on the module path. It seems pretty hard to
+  //      change our initialization order though: `new Run() -> Definitions.init -> PackageLoader.doComplete -> ClassPath.list`
+  //      happens before we even have a list of source files, which might include a module-info.java file, which is
+  //      relevant to the module graph construction.
+  private def indexLocations() {
+    for {
+      moduleLocation <- moduleLocations
+      location1 <- listLocationsForModules(fileManager, moduleLocation).asScala
+      location <- location1.asScala
+      path <- getLocationAsPaths(fileManager, location).asScala
+    } {
+      paths.addOne((path, moduleLocation, location))
+    }
+    val cpLocation = StandardLocation.CLASS_PATH
+    for (path <- getLocationAsPaths(fileManager, cpLocation).asScala)
+      paths.addOne((path, cpLocation, cpLocation))
+  }
+
+  private[nsc] def hasPackage(pkg: String): Boolean = {
+    throw new UnsupportedOperationException()
+  }
+
+  private[nsc] def packages(inPackage: String): Seq[PackageEntry] = {
+    val seen = mutable.Set[String]()
+    val result = immutable.ArraySeq.untagged.newBuilder[PackageEntry]
+    for {
+      (path, standardLocation, location) <- paths
+    } {
+      // We use NIO path API to list classpath/modulepath elements, rather than FileManager.list which
+      // only returns files, not sub-directories. The only way to find sub-directories with FileManager.list
+      // is to set `recurse = true`, which needlessly creates JavaFileObject and other metadata for the everything
+      // in the classpath.
+      val directory = path.resolve(inPackage.replaceAllLiterally(".", path.getFileSystem.getSeparator))
+      if (Files.isDirectory(directory)) {
+        val listing = Files.list(directory).iterator().asScala
+        for (file <- listing) {
+          if (Files.isDirectory(file)) {
+            val fileName = file.getFileName.toString
+            if (!seen(fileName) && !fileName.contains("-")) {
+              seen += fileName
+              val name = if (inPackage == "") fileName else inPackage + "." + fileName
+              result += PackageEntryImpl(name)
+            }
+          }
+        }
+      }
```

Comment:
We need an `else` branch for JARs on the module path, for which `path` is the `.jar` file.

I think the right approach is to index the package structure of the JARs eagerly, then use that index here.


WIP: https://github.com/retronym/scala/tree/review/jdk11

The indexing in that branch could be improved to be hierarchical.